libpod/pod.go

```diff
@@ -138,46 +103,109 @@ func (p *Pod) Namespace() string {
-		return nil
-	}
-	if resCopy != nil && resCopy.CPU != nil {
-		return resCopy
-	}
+	resCopy := &specs.Linux{}
 	empty := &specs.LinuxResources{
 		CPU: &specs.LinuxCPU{},
 	}
+	infra, err := p.runtime.GetContainer(p.state.InfraContainerID)
+	if err != nil {
+		return empty
+	}
+	conf, err := infra.specFromState()
+	if err != nil {
+		return empty
+	}
+	if err = JSONDeepCopy(conf.Linux, resCopy); err != nil {
+		return nil
+	}
+	if resCopy != nil && resCopy.Resources != nil && resCopy.Resources.CPU != nil {
+		return resCopy.Resources
+	}
+
 	return empty
 }
 
 // CPUPeriod returns the pod CPU period
 func (p *Pod) CPUPeriod() uint64 {
-	resCopy := &specs.LinuxResources{}
-	if err := JSONDeepCopy(p.config.InfraContainer.ResourceLimits, resCopy); err != nil {
+	resCopy := &specs.Spec{}
+	//resCopy.Linux = &specs.Linux{}
+	//resCopy.Linux.Resources = &specs.LinuxResources{}
+	if p.state.InfraContainerID == "" {
+		return 0
+	}
+	infra, err := p.runtime.GetContainer(p.state.InfraContainerID)
+	if err != nil {
+		return 0
+	}
+	conf := infra.Config()
+	if err := JSONDeepCopy(conf.Spec, resCopy); err != nil {
 		return 0
 	}
-	if resCopy != nil && resCopy.CPU != nil && resCopy.CPU.Period != nil {
-		return *resCopy.CPU.Period
+	if resCopy != nil && resCopy.Linux != nil && resCopy.Linux.Resources != nil && resCopy.Linux.Resources.CPU != nil && resCopy.Linux.Resources.CPU.Period != nil {
+		return *resCopy.Linux.Resources.CPU.Period
 	}
 	return 0
 }
 
 // CPUQuota returns the pod CPU quota
 func (p *Pod) CPUQuota() int64 {
-	resCopy := &specs.LinuxResources{}
-	if err := JSONDeepCopy(p.config.InfraContainer.ResourceLimits, resCopy); err != nil {
+	resCopy := &specs.Spec{}
+	infra, err := p.runtime.GetContainer(p.state.InfraContainerID)
+	if err != nil {
 		return 0
 	}
-	if resCopy != nil && resCopy.CPU != nil && resCopy.CPU.Quota != nil {
-		return *resCopy.CPU.Quota
+	conf := infra.Config()
+	if err := JSONDeepCopy(conf.Spec, resCopy); err != nil {
+		return 0
+	}
+	if resCopy != nil && resCopy.Linux != nil && resCopy.Linux.Resources != nil && resCopy.Linux.Resources.CPU != nil && resCopy.Linux.Resources.CPU.Quota != nil {
+		return *resCopy.Linux.Resources.CPU.Quota
 	}
 	return 0
 }
 
 // PidMode returns the PID mode given by the user ex: pod, private...
 func (p *Pod) PidMode() string {
-	return string(p.config.InfraContainer.PidNS.NSMode)
+	infra, err := p.runtime.GetContainer(p.state.InfraContainerID)
+	if err != nil {
+		return ""
+	}
+	conf := infra.Config()
+	ctrSpec := conf.Spec
+	if ctrSpec != nil && ctrSpec.Linux != nil {
+		for _, ns := range ctrSpec.Linux.Namespaces {
+			if ns.Type == specs.PIDNamespace {
+				if ns.Path != "" {
+					return fmt.Sprintf("ns:%s", ns.Path)
+				}
+				return "private"
+			}
+		}
+		return "host"
+	}
+	return ""
+}
+
+// PidMode returns the PID mode given by the user ex: pod, private...
+func (p *Pod) UserMode() string {
+	infra, err := p.runtime.GetContainer(p.state.InfraContainerID)
```

Comment:
You should make a helper for this - `p.InfraContainer()` (this might already exist)

Reply:
I originally used this but `.InfraContainer` acquires the pod lock and caused the program to freeze upon inspect

Reply:
You can make an unlocked version (we usually call it infraContainer() so it's not public). Usually, `InfraContainer()` will just acquire the lock, then call `infraContainer()`, which does the work